### ISSUE TITLE
NAT-01-008 node-nats: DoSe via malicious JSON structure (low)

### ISSUE THREAD

**Issue by aricart** [nats-io/node-nats#251]
The NodeJS client can define that any message payloads are converted from/to JSON.
This is achieved by setting `json` parameter to `true` in the `connect` function. It was
discovered that a possibility to trigger an exception nevertheless exists and can lead to a
crash of the client. This is triggered by sending an object which overwrites the toString
function with a string. As soon as the received message is used as a string in any
context, an exception is triggered.

**Comment by aricart** [comment]
JSON doesn’t allow code injection. However, in NAT-01-008 particular case, the issue is that a well-known method of all objects (“toString”) is usurped as a property with a value that is NOT a function (string, number, array, object, boolean, null). 

If user code tries to use the value by appending such as appending with the `+` operator, a `Cannot convert object to primitive value` is raised.
Here’s a snippet for the issue in question:

```javascript
var a = {hello: a};
a.toString = "hi";
console.log(a); // { hello: undefined, toString: 'hi' }
var o = JSON.parse(JSON.stringify(a))
console.log(o); // { toString: 'hi' }
console.log(o.toString); // ‘hi’
console.log(o + "a"); // this is bad - value cannot be converted
o.toString(); // toString is not a function)
```

This error is the correct/expected behavior for any unguarded access to any value that is unknown or has an unknown origin. Safe programming practices for data of an unknown or untrusted origin would require the client to test the type of property that it is to ensure safety before it is accessed.
Not only that, but this also points out that even if the issue was addressed in some way, the problem would exist for any field in any untrusted JSON (or payload for that matter). 

For example, in a different scenario, if a JSON object has a property expected to be a number but that contains a string, and user code tries to use the value in a context where a number is required, a similar type of error would be raised.

We believe that this issue does not apply to node-nats for a number of reasons:

- Encoding/Decoding JSON does the correct thing, and injection is impossible.
- Message rejection is not possible without inspecting all properties for all objects that are parsed from JSON. 
- Removing properties which we think are malicious is outside of the scope of the library
- We don’t have knowledge of all JSON structures regardless of however ill-formed they are - what if that ‘toString’ property is actually valid and a string representation that can be presented properly

For the above reasons, we feel this issue doesn’t necessarily need to be addressed save to say that programs handling foreign input should check their input prior to accessing it is always a good practice.  We have added the following documentation https://github.com/nats-io/node-nats#json.
